File: array/problem39/problem39.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        result: List[List[int]] = []

        def backtrack(start: int, remaining: int, path: List[int]) -> None:
            if remaining == 0:
                result.append(path.copy())
                return
            for i in range(start, len(candidates)):
                c = candidates[i]
                if c > remaining:
                    break
                path.append(c)
                backtrack(i, remaining - c, path)
                path.pop()

        backtrack(0, target, [])
        return result

File: array/problem39/problem39.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
    void backtrack(
        vector<int>& candidates, int start, int remaining,
        vector<int>& path, vector<vector<int>>& result
    ) {
        if (remaining == 0) {
            result.push_back(path);
            return;
        }
        for (int i = start; i < static_cast<int>(candidates.size()); ++i) {
            int c = candidates[i];
            if (c > remaining) break;
            path.push_back(c);
            backtrack(candidates, i, remaining - c, path, result);
            path.pop_back();
        }
    }

public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> result;
        vector<int> path;
        backtrack(candidates, 0, target, path, result);
        return result;
    }
};
